staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,404 @@
+	result := &cachingObject{}
+	func() {
+		o.lock.RLock()
+		defer o.lock.RUnlock()
+		result.object = o.object.DeepCopyObject().(metaRuntimeInterface)
+	}()
+	result.serializations.Store(make(serializationsCache))
+	return result
+}
+
+func (o *cachingObject) invalidateCacheLocked() {
+	if cache, ok := o.serializations.Load().(serializationsCache); ok && len(cache) == 0 {
+		return
+	}
+	klog.Warningf("Unexpected cache invalidation for %#v", o.object)
+	o.serializations.Store(make(serializationsCache))
+}
+
+// The following functions implement metav1.Object interface:
+// - getters simply delegate for underlying object
+// - SetSelfLink stores the value passed in the first call
+// - all other setters ignore the passed value.
+
+func (o *cachingObject) GetNamespace() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetNamespace()
+}
+func (o *cachingObject) SetNamespace(namespace string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetNamespace() == namespace {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetNamespace(namespace)
+}
+func (o *cachingObject) GetName() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetName()
+}
+func (o *cachingObject) SetName(name string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetName() == name {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetName(name)
+}
+func (o *cachingObject) GetGenerateName() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetGenerateName()
+}
+func (o *cachingObject) SetGenerateName(name string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetGenerateName() == name {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetGenerateName(name)
+}
+func (o *cachingObject) GetUID() types.UID {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetUID()
+}
+func (o *cachingObject) SetUID(uid types.UID) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetUID() == uid {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetUID(uid)
+}
+func (o *cachingObject) GetResourceVersion() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetResourceVersion()
+}
+func (o *cachingObject) SetResourceVersion(version string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetResourceVersion() == version {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetResourceVersion(version)
+}
+func (o *cachingObject) GetGeneration() int64 {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetGeneration()
+}
+func (o *cachingObject) SetGeneration(generation int64) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetGeneration() == generation {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetGeneration(generation)
+}
+func (o *cachingObject) GetSelfLink() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetSelfLink()
+}
+func (o *cachingObject) SetSelfLink(selfLink string) {
+	o.lock.Lock()
```

Comment:
consider doing an RLock first to check if we need to invalidate and actually set. maybe benchmark to see if the added code complexity is worth it

Reply:
might consider refactoring these to call a helper to manage the RLock/Unlock/invalidateCacheLocked calls

```go
o.conditionalSet(
  func()bool{ return o.object.GetSelfLink() == selfLink},
  func(){ o.object.SetSelfLink(selfLink)},
)
```

Reply:
done